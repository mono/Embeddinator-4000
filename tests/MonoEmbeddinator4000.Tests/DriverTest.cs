using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using CppSharp;
using CppSharp.Generators;
using NUnit.Framework;

namespace Embeddinator.Tests
{
    /// <summary>
    /// A set of integration tests / approval tests verifying generated java code
    /// </summary>
    [TestFixture]
    public class DriverTest : TempFileTest
    {
        string outputDir;
        Project project;
        Options options;
        Driver driver;

        [SetUp]
        public override void SetUp()
        {
            base.SetUp();

            outputDir = Path.Combine(Path.GetDirectoryName(GetType().Assembly.Location), "output");
            if (!Directory.Exists(outputDir))
                Directory.CreateDirectory(outputDir);

            temp = Path.Combine(outputDir, "hello.dll");
            tempFiles = new List<string> { temp, outputDir };

            project = new Project();
            options = new Options
            {
                GeneratorKind = GeneratorKind.Java,
                CompileCode = true,
            };

            options.OutputDir =
                project.OutputPath = outputDir;
        }

        void RunDriver(string resourceFile, CompilerParameters parameters = null)
        {
            if (parameters == null)
            {
                parameters = new CompilerParameters();
            }
            parameters.OutputAssembly = temp;
            AssemblyGenerator.CreateFromResource(resourceFile, parameters);

            project.Assemblies.Insert(0, temp);

            driver = new Driver(project, options);
            Assert.IsTrue(driver.Run(), "Call to Driver.Run() failed!");
        }

        [Test]
        public void Empty()
        {
            driver = new Driver(new Project(), new Options
            {
                GeneratorKind = GeneratorKind.Java,
            });
            Assert.IsTrue(driver.Run()); //This runs but doesn't throw
        }

        [Test]
        public void HelloFiles()
        {
            RunDriver("Hello");

            var builder = new StringBuilder();
            foreach (var file in driver.Output.Files.Keys)
            {
                //NOTE: replace \ so this works on Windows
                builder.AppendLine(file.Replace('\\', '/'));
            }
            Approvals.Verify(builder.ToString());
        }

        [Test]
        public void Hello_java()
        {
            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path);
        }

        [Test]
        public void Native_hello_dll_java()
        {
            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.Last());
            Approvals.VerifyFile(path);
        }

        [Test]
        public void UpperCase_java()
        {
            RunDriver("HelloUpper");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path); //TODO: I don't know if "String wORLD()" is what we want
        }

        [Test]
        public void AssemblyWithDots()
        {
            temp = Path.Combine(outputDir, "hello.with.dots.dll");
            tempFiles.Add(temp);

            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path);
        }

        [Test]
        public void AssemblyWithUpperCase()
        {
            temp = Path.Combine(outputDir, "HELLO.dll");
            tempFiles.Add(temp);

            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path);
        }

        [Test, Category("Slow"), Platform("MacOSX")]
        public void JarFileContents()
        {
            options.Compilation.Platform = TargetPlatform.MacOS;
            options.GeneratorKind = GeneratorKind.C;
            RunDriver("Hello");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello");

            var aar = Path.Combine(options.OutputDir, "Hello.jar");
            Approvals.VerifyZipFile(aar);
        }

        [Test, Category("Slow")]
        public void AarFileContents()
        {
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            RunDriver("Hello");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello");

            var aar = Path.Combine(options.OutputDir, "Hello.aar");
            Approvals.VerifyZipFile(aar);
        }

        [Test, Category("Slow")]
        public void AarFileContentsDebug()
        {
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            options.Compilation.DebugMode = true;
            RunDriver("Hello");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello");

            var aar = Path.Combine(options.OutputDir, "Hello.aar");
            Approvals.VerifyZipFile(aar);
        }

        /// <summary>
        /// NOTE: C and Java generators were failing on a subclass of EventArgs due to EventArgs.Empty
        /// </summary>
        [Test, Category("Slow")]
        public void EventArgsEmpty()
        {
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            options.Compilation.DebugMode = true;
            RunDriver("EventArgsEmpty");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("EventArgsEmpty");
        }

        [Test, Category("Slow")]
        public void Enums()
        {
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            options.Compilation.DebugMode = true;
            RunDriver("Enums");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Enums");
        }

        [Test, Category("Slow")]
        public void Interfaces()
        {
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            options.Compilation.DebugMode = true;
            RunDriver("Interfaces");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Interfaces");
        }

        /// <summary>
        /// Validates we get native libraries from the assembly
        /// </summary>
        [Test, Category("Slow")]
        public void AndroidNativeLibraries()
        {
            var nativeLibrariesZip = Path.Combine(Path.GetDirectoryName(GetType().Assembly.Location), "..", "..", "Samples", "__AndroidNativeLibraries__.zip");
            var parameters = new CompilerParameters();
            parameters.EmbeddedResources.Add(nativeLibrariesZip);

            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            RunDriver("Hello", parameters);
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello", parameters);

            var aar = Path.Combine(options.OutputDir, "Hello.aar");
            Approvals.VerifyZipFile(aar);
        }

        /// <summary>
        /// Validates we get resources/assets from the assembly
        /// </summary>
        [Test, Category("Slow")]
        public void AndroidResources()
        {
            var libraryProjectsZip = Path.Combine(Path.GetDirectoryName(GetType().Assembly.Location), "..", "..", "Samples", "__AndroidLibraryProjects__.zip");
            var parameters = new CompilerParameters();
            parameters.EmbeddedResources.Add(libraryProjectsZip);

            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            RunDriver("Hello", parameters);
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello", parameters);

            var aar = Path.Combine(options.OutputDir, "Hello.aar");
            Approvals.VerifyZipFile(aar);
        }

        /// <summary>
        /// Validates we get native libraries and resources/assets from a dependency
        /// NOTE: the dependent assembly should be passed as an input assembly to E4K
        /// </summary>
        [Test, Category("Slow")]
        public void AndroidDependencies()
        {
            var dependency = Path.Combine(outputDir, "dependency.dll");
            var nativeLibrariesZip = Path.Combine(Path.GetDirectoryName(GetType().Assembly.Location), "..", "..", "Samples", "__AndroidNativeLibraries__.zip");
            var libraryProjectsZip = Path.Combine(Path.GetDirectoryName(GetType().Assembly.Location), "..", "..", "Samples", "__AndroidLibraryProjects__.zip");
            var parameters = new CompilerParameters();
            parameters.OutputAssembly = dependency;
            parameters.EmbeddedResources.Add(nativeLibrariesZip);
            parameters.EmbeddedResources.Add(libraryProjectsZip);
            AssemblyGenerator.CreateFromResource("HelloUpper", parameters);
            project.Assemblies.Add(dependency);

            parameters = new CompilerParameters();
            parameters.ReferencedAssemblies.Add(dependency);
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            RunDriver("Hello", parameters);
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello", parameters);

            var aar = Path.Combine(options.OutputDir, "Hello.aar");
            Approvals.VerifyZipFile(aar);
        }
    }
}
