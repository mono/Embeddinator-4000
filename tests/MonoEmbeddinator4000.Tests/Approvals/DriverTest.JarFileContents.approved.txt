META-INF/
META-INF/MANIFEST.MF
assemblies/
assemblies/hello.dll
assemblies/mscorlib.dll
com/
com/sun/
com/sun/jna/
com/sun/jna/aix-ppc/
com/sun/jna/aix-ppc/libjnidispatch.a
com/sun/jna/aix-ppc64/
com/sun/jna/aix-ppc64/libjnidispatch.a
com/sun/jna/AltCallingConvention.class
com/sun/jna/Callback$UncaughtExceptionHandler.class
com/sun/jna/Callback.class
com/sun/jna/CallbackParameterContext.class
com/sun/jna/CallbackProxy.class
com/sun/jna/CallbackReference$AttachOptions.class
com/sun/jna/CallbackReference$DefaultCallbackProxy.class
com/sun/jna/CallbackReference$NativeFunctionHandler.class
com/sun/jna/CallbackReference.class
com/sun/jna/CallbackResultContext.class
com/sun/jna/CallbackThreadInitializer.class
com/sun/jna/darwin/
com/sun/jna/darwin/libjnidispatch.jnilib
com/sun/jna/DefaultTypeMapper$Entry.class
com/sun/jna/DefaultTypeMapper.class
com/sun/jna/ELFAnalyser.class
com/sun/jna/FromNativeContext.class
com/sun/jna/FromNativeConverter.class
com/sun/jna/Function$NativeMappedArray.class
com/sun/jna/Function$PointerArray.class
com/sun/jna/Function$PostCallRead.class
com/sun/jna/Function.class
com/sun/jna/FunctionMapper.class
com/sun/jna/FunctionParameterContext.class
com/sun/jna/FunctionResultContext.class
com/sun/jna/IntegerType.class
com/sun/jna/InvocationMapper.class
com/sun/jna/LastErrorException.class
com/sun/jna/Library$Handler$FunctionInfo.class
com/sun/jna/Library$Handler.class
com/sun/jna/Library.class
com/sun/jna/linux-aarch64/
com/sun/jna/linux-aarch64/libjnidispatch.so
com/sun/jna/linux-arm/
com/sun/jna/linux-arm/libjnidispatch.so
com/sun/jna/linux-armel/
com/sun/jna/linux-armel/libjnidispatch.so
com/sun/jna/linux-ppc/
com/sun/jna/linux-ppc/libjnidispatch.so
com/sun/jna/linux-ppc64le/
com/sun/jna/linux-ppc64le/libjnidispatch.so
com/sun/jna/linux-x86/
com/sun/jna/linux-x86/libjnidispatch.so
com/sun/jna/linux-x86-64/
com/sun/jna/linux-x86-64/libjnidispatch.so
com/sun/jna/Memory$SharedMemory.class
com/sun/jna/Memory.class
com/sun/jna/MethodParameterContext.class
com/sun/jna/MethodResultContext.class
com/sun/jna/Native$1.class
com/sun/jna/Native$2.class
com/sun/jna/Native$3.class
com/sun/jna/Native$4.class
com/sun/jna/Native$5.class
com/sun/jna/Native$6.class
com/sun/jna/Native$7.class
com/sun/jna/Native$AWT.class
com/sun/jna/Native$Buffers.class
com/sun/jna/Native$ffi_callback.class
com/sun/jna/Native.class
com/sun/jna/NativeLibrary$1.class
com/sun/jna/NativeLibrary$2.class
com/sun/jna/NativeLibrary.class
com/sun/jna/NativeLong.class
com/sun/jna/NativeMapped.class
com/sun/jna/NativeMappedConverter.class
com/sun/jna/NativeString$StringMemory.class
com/sun/jna/NativeString.class
com/sun/jna/Platform.class
com/sun/jna/Pointer$1.class
com/sun/jna/Pointer$Opaque.class
com/sun/jna/Pointer.class
com/sun/jna/PointerType.class
com/sun/jna/ptr/
com/sun/jna/ptr/ByReference.class
com/sun/jna/ptr/ByteByReference.class
com/sun/jna/ptr/DoubleByReference.class
com/sun/jna/ptr/FloatByReference.class
com/sun/jna/ptr/IntByReference.class
com/sun/jna/ptr/LongByReference.class
com/sun/jna/ptr/NativeLongByReference.class
com/sun/jna/ptr/PointerByReference.class
com/sun/jna/ptr/ShortByReference.class
com/sun/jna/StringArray.class
com/sun/jna/Structure$1.class
com/sun/jna/Structure$2.class
com/sun/jna/Structure$3.class
com/sun/jna/Structure$AutoAllocated.class
com/sun/jna/Structure$ByReference.class
com/sun/jna/Structure$ByValue.class
com/sun/jna/Structure$FFIType$FFITypes.class
com/sun/jna/Structure$FFIType$size_t.class
com/sun/jna/Structure$FFIType.class
com/sun/jna/Structure$LayoutInfo.class
com/sun/jna/Structure$StructField.class
com/sun/jna/Structure$StructureSet.class
com/sun/jna/Structure.class
com/sun/jna/StructureReadContext.class
com/sun/jna/StructureWriteContext.class
com/sun/jna/sunos-x86-64/
com/sun/jna/sunos-x86-64/libjnidispatch.so
com/sun/jna/ToNativeContext.class
com/sun/jna/ToNativeConverter.class
com/sun/jna/TypeConverter.class
com/sun/jna/TypeMapper.class
com/sun/jna/Union.class
com/sun/jna/VarArgsChecker$1.class
com/sun/jna/VarArgsChecker$NoVarArgsChecker.class
com/sun/jna/VarArgsChecker$RealVarArgsChecker.class
com/sun/jna/VarArgsChecker.class
com/sun/jna/Version.class
com/sun/jna/WeakMemoryHolder.class
com/sun/jna/win32/
com/sun/jna/win32/DLLCallback.class
com/sun/jna/win32/StdCall.class
com/sun/jna/win32/StdCallFunctionMapper.class
com/sun/jna/win32/StdCallLibrary$StdCallCallback.class
com/sun/jna/win32/StdCallLibrary.class
com/sun/jna/win32/W32APIFunctionMapper.class
com/sun/jna/win32/W32APIOptions$1.class
com/sun/jna/win32/W32APIOptions$2.class
com/sun/jna/win32/W32APIOptions.class
com/sun/jna/win32/W32APITypeMapper$1.class
com/sun/jna/win32/W32APITypeMapper$2.class
com/sun/jna/win32/W32APITypeMapper.class
com/sun/jna/win32-x86/
com/sun/jna/win32-x86/jnidispatch.dll
com/sun/jna/win32-x86-64/
com/sun/jna/win32-x86-64/jnidispatch.dll
com/sun/jna/WString.class
darwin/
darwin/libhello.dylib
darwin/libmonosgen-2.0.dylib
hello/
hello/example/
hello/example/Hello.class
hello/Native_hello.class
mono/
mono/embeddinator/
mono/embeddinator/Decimal$ByReference.class
mono/embeddinator/Decimal$ByValue.class
mono/embeddinator/Decimal.class
mono/embeddinator/DesktopImpl.class
mono/embeddinator/GString.class
mono/embeddinator/NullRefParameterException.class
mono/embeddinator/Obj.class
mono/embeddinator/Out.class
mono/embeddinator/Ref.class
mono/embeddinator/Runtime$1.class
mono/embeddinator/Runtime$RuntimeLibrary$Error$ByValue.class
mono/embeddinator/Runtime$RuntimeLibrary$Error.class
mono/embeddinator/Runtime$RuntimeLibrary$ErrorCallback.class
mono/embeddinator/Runtime$RuntimeLibrary$ErrorType.class
mono/embeddinator/Runtime$RuntimeLibrary$GString$ByReference.class
mono/embeddinator/Runtime$RuntimeLibrary$GString$ByValue.class
mono/embeddinator/Runtime$RuntimeLibrary$GString.class
mono/embeddinator/Runtime$RuntimeLibrary$MonoDecimal.class
mono/embeddinator/Runtime$RuntimeLibrary.class
mono/embeddinator/Runtime.class
mono/embeddinator/RuntimeException.class
mono/embeddinator/UnsignedByte.class
mono/embeddinator/UnsignedInt.class
mono/embeddinator/UnsignedLong.class
mono/embeddinator/UnsignedShort.class
mono/embeddinator/Utilities.class
